File: evernote_client/mcp/write_tools.py
"""Write MCP tools for Evernote."""

from __future__ import annotations

from evernote_client import service
from evernote_client.models import CreatedNote, NoteMetadata

from .app import mcp


@mcp.tool()
def create_note(
    title: str,
    content: str,
    notebook_name: str = "",
    tags: list[str] | None = None,
) -> CreatedNote:
    """Create a new note with Markdown content.

    Args:
        title: Note title
        content: Note content in Markdown format
        notebook_name: Target notebook name (uses default if empty)
        tags: List of tag names to apply
    """
    return service.create_note(
        title=title,
        content=content,
        notebook_name=notebook_name,
        tags=tags,
    )


@mcp.tool()
def tag_note(guid: str, tags: list[str]) -> NoteMetadata:
    """Add tags to an existing note.

    Args:
        guid: Note GUID
        tags: Tag names to add
    """
    return service.tag_note(guid, tags)


@mcp.tool()
def untag_note(guid: str, tags: list[str]) -> NoteMetadata:
    """Remove tags from an existing note.

    Args:
        guid: Note GUID
        tags: Tag names to remove
    """
    return service.untag_note(guid, tags)


@mcp.tool()
def move_note(guid: str, notebook_name: str) -> NoteMetadata:
    """Move a note to a different notebook.

    Args:
        guid: Note GUID
        notebook_name: Target notebook name
    """
    return service.move_note(guid, notebook_name)

File: src/server/write-tools.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod/v4';
import * as service from '../service.js';
import { jsonResult } from './response.js';

export function registerWriteTools(server: McpServer): void {
  server.tool(
    'create_note',
    'Create a new note with Markdown content',
    {
      title: z.string().describe('Note title'),
      content: z
        .string()
        .optional()
        .describe('Note content in Markdown format'),
      notebook_name: z
        .string()
        .optional()
        .describe('Target notebook name (uses default if empty)'),
      tags: z
        .array(z.string())
        .optional()
        .describe('List of tag names to apply'),
    },
    async (args) =>
      jsonResult(
        await service.createNote(
          args.title,
          args.content ?? '',
          args.notebook_name ?? '',
          args.tags ?? null
        )
      )
  );

  server.tool(
    'tag_note',
    'Add tags to an existing note',
    {
      guid: z.string().describe('Note GUID'),
      tags: z.array(z.string()).describe('Tag names to add'),
    },
    async (args) => jsonResult(await service.tagNote(args.guid, args.tags))
  );

  server.tool(
    'untag_note',
    'Remove tags from an existing note',
    {
      guid: z.string().describe('Note GUID'),
      tags: z.array(z.string()).describe('Tag names to remove'),
    },
    async (args) => jsonResult(await service.untagNote(args.guid, args.tags))
  );

  server.tool(
    'move_note',
    'Move a note to a different notebook',
    {
      guid: z.string().describe('Note GUID'),
      notebook_name: z.string().describe('Target notebook name'),
    },
    async (args) =>
      jsonResult(await service.moveNote(args.guid, args.notebook_name))
  );
}
